File: backend-csharp/BioPagesController.cs
using BackendCsharp.Contracts;
using BackendCsharp.Services;
using Microsoft.AspNetCore.Mvc;

namespace BackendCsharp;

[ApiController]
[Route("bio-pages")]
public class BioPagesController(BioPagesService bioPagesService) : ControllerBase
{
    [HttpGet]
    public IActionResult FindAll()
    {
        return Ok(bioPagesService.FindAll());
    }

    [HttpGet("{id}")]
    public IActionResult FindOne(string id)
    {
        var bioPage = bioPagesService.FindOne(id);
        return bioPage is null ? NotFound(new { message = "Bio page not found" }) : Ok(bioPage);
    }

    [HttpGet("handle/{handle}")]
    public IActionResult FindOneByHandle(string handle)
    {
        var bioPage = bioPagesService.FindOneByHandle(handle);
        return bioPage is null ? NotFound(new { message = "Bio page not found" }) : Ok(bioPage);
    }

    [HttpPost]
    public IActionResult Create([FromBody] CreateBioPageRequest payload)
    {
        if (!BioPagesService.IsValidCreatePayload(payload))
        {
            return BadRequest(new { message = "Invalid payload" });
        }

        try
        {
            var created = bioPagesService.Create(payload);
            return Ok(created);
        }
        catch (InvalidOperationException ex)
        {
            return Conflict(new { message = ex.Message });
        }
    }

    [HttpPatch("{id}")]
    public IActionResult Update(string id, [FromBody] UpdateBioPageRequest payload)
    {
        if (!BioPagesService.IsValidUpdatePayload(payload))
        {
            return BadRequest(new { message = "Invalid payload" });
        }

        try
        {
            var updated = bioPagesService.Update(id, payload);
            return updated is null ? NotFound(new { message = "Bio page not found" }) : Ok(updated);
        }
        catch (InvalidOperationException ex)
        {
            return Conflict(new { message = ex.Message });
        }
    }
}

File: backend-ts/src/bio-pages/bio-pages.controller.ts
import { Body, Controller, Get, Param, Patch, Post } from '@nestjs/common';
import { BioPagesService } from './bio-pages.service';
import { CreateBioPageDto } from './dto/create-bio-page.dto';
import { UpdateBioPageDto } from './dto/update-bio-page.dto';

@Controller('bio-pages')
export class BioPagesController {
  constructor(private readonly bioPagesService: BioPagesService) {}

  @Get()
  findAll() {
    return this.bioPagesService.findAll();
  }

  @Get(':id')
  findOne(@Param('id') id: string) {
    return this.bioPagesService.findOne(id);
  }

  @Get('handle/:handle')
  findOneByHandle(@Param('handle') handle: string) {
    return this.bioPagesService.findOneByHandle(handle);
  }

  @Post()
  create(@Body() payload: CreateBioPageDto) {
    return this.bioPagesService.create(payload);
  }

  @Patch(':id')
  update(@Param('id') id: string, @Body() payload: UpdateBioPageDto) {
    return this.bioPagesService.update(id, payload);
  }
}
